sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -633,7 +633,9 @@ public Symbol visitArrayComparisonExpression(ArrayComparisonExpression node, Exp
             // int_col = ANY([1, 2, 3])     -> must cast to int instead of long (otherwise lucene query would be slow)
             // null = ANY([1, 2])           -> must not cast to null
             if (SymbolVisitors.any(symbol -> symbol instanceof Field, leftSymbol) || rightInnerType == DataTypes.UNDEFINED) {
-                arraySymbol = castIfNeededOrFail(arraySymbol, new ArrayType(leftSymbol.valueType()));
+                CollectionType newArrayValueType = (CollectionType) DataTypes.fromId(arraySymbolType.id());
```

Comment:
I did consider it but it requires slightly bigger changes. If we add the factory method to CollectionType, we still have to cast.